docs/proposals/server-side-apply-field-managers.md

```diff
@@ -0,0 +1,123 @@
+# Server-Side Apply with Unique Field Managers
+
+This proposal is a follow up on the [original Server-Side Apply proposal](./server-side-apply.md), and seeks to make Argo more flexible and give users the ability to apply changes to shared Kuberentes resources at a granular level. It also considers how to remove field-level changes when an Application is destroyed.
+
+To quote the [Kubernetes docs][1]:
+
+> Server-Side Apply helps users and controllers manage their resources through declarative configuration. Clients can create and modify objects declaratively by submitting their _fully specified intent_.
+
+> A fully specified intent is a partial object that only includes the fields and values for which the user has an opinion. That intent either creates a new object (using default values for unspecified fields), or is combined, by the API server, with the existing object.
+
+## Example
+
+Consider the case where you're working on a new feature for your app. You'd like to review these changes without disturbing your staging environment, so you use an `ApplicationSet` with a pull request generator to create [review apps][3] dynamically. These review apps configure a central Consul `IngressGateway`. Each review app needs to add a service to the `IngressGateway` upon creation, and remove that service upon deletion.
+
+```yaml
+apiVersion: consul.hashicorp.com/v1alpha1
+kind: IngressGateway
+metadata:
+  name: review-app-ingress-gateway
+  namespace: default
+spec:
+  listeners:
+  - port: 443
+    protocol: http
+    services:
+    - name: review-app-1
+      namespace: my-cool-app
+    - name: review-app-2
+      namespace: my-groovy-app
+    - name: review-app-3
+      namespace: my-incredible-app
+```
+
+---
+
+## Open Questions
+
+### [Q-1] What should the behavior be for a server-side applied resource upon Application deletion?
+
+The current behavior is to delete all objects "owned" by the Application. A user could choose to leave the resource behind with `Prune=false`, but that would also leave behind any config added to the shared object. Should the default delete behavior for server-side applied resources be to remove any fields that match that Application's [field manager][2]?
+
+### [Q-2] What sync status should the UI display on a shared resource?
+
+If an Application has successfully applied its partial spec to a shared resource, should it display as "in sync"? Or should it show "out of sync" when there are other changes to the shared object?
```

Comment:
This will vary depending on the diff strategy being used by the Application. The current diff logic (for non server-side apply sync) only takes into account the live state, the git state and the `last-applied-configuration` annotation. It executes a 3-way diff between the 3 states to calculate the diff. The new server-side diff strategy takes into consideration the manager instead to define field ownership. We would have to validate how server-side diff would behave with dynamic managers.